### DISCUSSION TITLE
Unified Create from Template Experience

### DISCUSSION THREAD

**Discussion by ntotten** [forcedotcom/salesforcedx-vscode#5738] [Ideas]
Today, we provide individual commands for creating each different kind of template as well as right click menu items for every create experience. This has started to become unwieldy for several reasons:

1) The right click menu is growing long
2) The SFDX command list is growing long
3) The right-click menu cannot be filtered by folder type so you end up with things like `SFDX: Create Visual Force Page` on every folder in the entire project.

![image](https://user-images.githubusercontent.com/282782/51754984-9db5ef80-208b-11e9-9450-b1397d4b4e48.png)

![image](https://user-images.githubusercontent.com/282782/51755002-a7d7ee00-208b-11e9-9183-afa9b122df56.png)

The proposed solution is to simplify the creation of any source into a single command that provides options. The command would be `SFDX: Create From Template`. It would be exposed in the command pallet and the right-click menu on all folders in a project.

When a user runs the command by right-clicking a folder, we should have a context as to where that command originated. As such we will present a list of templates to the user based on the folder selected. We will use the *default* folder locations of metadata to filter that list. So for example, if they run the command from the `classes` folder they would be presented with a list of apex class templates to create.

The UI to pick classes would initially be the built-in visual studio picker UI, but could eventually evolve into a more advanced custom template UI.

From an implementation standpoint this would also mean that each salesforce extension would no longer expose its own create commands. We would have a single create command exposed from the core extension. Thus ensuring a completely consistent experience no matter which type of source is created.

**Comment by vazexqi** [comment]
> We will use the default folder locations of metadata to filter that list. So for example, if they run the command from the classes folder they would be presented with a list of apex class templates to create.

*Minor*

1. What do you do when the user doesn't follow the names, which is allowed for some types since we use the file extension? Do we just fall back and present all of it?
1. What do you do when you are on an Aura, LWC, or components that support subfolders? The name of the component can be arbitrary and have no mapping.  Do we just fall back and present all of it?

I think we were asking for an VS Code API to at least get the full path to the resource but that was `never`ed. The full path might be able to help us here.

**Comment by ntotten** [comment]
> What do you do when the user doesn't follow the names, which is allowed for some types since we use the file extension? Do we just fall back and present all of it?

We wouldn't show anything. The assumption here is that if a user is using custom names (which is technically allowed, but really doesnt work that well in VS Code) they are an advanced user and will use the command+p menu item which will show all the template and then ask them where they want to create it.

> What do you do when you are on an Aura, LWC, or components that support subfolders? The name of the component can be arbitrary and have no mapping. Do we just fall back and present all of it?

I think we can show the aura components on any subfolder.

> I think we were asking for an VS Code API to at least get the full path to the resource but that was nevered. The full path might be able to help us here.

yes, this is the solution to that nevered issue which they wont do. That issue would allow us to filter when a command is shown in the tree, but that isnt needed anymore with this solution. The command would be a single command and always the same, but once the command is run we do know the path so we can filter the templates that are shown based on that path context.

**Comment by ntotten** [comment]
See: https://github.com/visualzoran/vz-file-templates

[a comment by an automated account is not shown]
